### PR TITLE
feat(datalist): add datalist component

### LINKED ISSUE #3813: Datalist component
## Summary

Implement an HTML `datalist` element as an Angular control, and make sure it fits well within our forms layouts and styling.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist 

* What is the change? This is a new component but of an existing HTML element that is widely supported
* Why should it go in Clarity? It provides a limited combobox like behavior and might be ideal in many scenarios
* Does this change impact existing behaviors? If so how? No its new
* If this change introduces a new behavior, is this behavior accessible? It should be the same as other form controls

## Use case

We've seen people using dropdowns to create a combobox like functionality, but a datalist is often sufficient. 

## Examples

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist

## Workarounds

_Is there a workaround for this feature at the moment, even if it is less than ideal?_

### PR DESCRIPTION
**Pull request by hippee-lee**
NOTE: I'll add website docs in a second commit on this PR. 

- Adds the ClrDatalist component for Clarity forms
- closes #3813
- adopts and closes #3895 

Signed-off-by: Matt Hippely <mhippely@vmware.com>
Signed-off-by: Marcu Alexandru Stefanel <alexandru.marcu@bytex.ro>

## PR Checklist

Please check if your PR fulfills the following requirements:

* [X] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?
This change adds the ClrDatalist component that builds on top of the native datalist element and allows it to work with forms. 

* [ ] Bugfix
* [x] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [x] Documentation content changes
* [ ] clarity.design website / infrastructure changes
* [ ] Other... Please describe:

## What is the current behavior?
There is no form support for datalist elements

Issue Number: N/A

## What is the new behavior?
Forms can use datalist elements
## Does this PR introduce a breaking change?

* [ ] Yes
* [x] No

### CODE REVIEW

File: src/clr-angular/forms/tests/control.spec.ts [hunk 1]
```diff
@@ -66,6 +75,8 @@ function fullTest(description, testContainer, testControl, testComponent, contro
       ifErrorService = control.injector.get(IfErrorService);
       markControlService = control.injector.get(MarkControlService);
       layoutService = control.injector.get(LayoutService);
+      controlIdService = control.injector.get(ControlIdService);
```
Comment (bdryanovski) [inline src/clr-angular/forms/tests/control.spec.ts]: both `controlIdService` and `datalistIdService` are assigned but never used

File: src/clr-angular/forms/datalist/datalist.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@Directive({
+  selector: 'datalist',
+  host: {
+    '[id]': 'datalistId',
+  },
+})
+export class ClrDatalist implements AfterContentInit {
+  private subscriptions: Subscription[] = [];
+  constructor(private datalistIdService: DatalistIdService) {}
+  datalistId;
```
Comment (bdryanovski) [inline src/clr-angular/forms/datalist/datalist.ts]: `datalistId` is missing the type definition

File: src/clr-angular/wizard/wizard.spec.ts [hunk 3]
```diff
@@ -29,6 +29,7 @@ export default function(): void {
           wizard = context.clarityDirective;
           component = context.hostComponent;
           context.detectChanges();
+          context.detectChanges();
```
Comment (bdryanovski) [inline src/clr-angular/wizard/wizard.spec.ts]: I manage to pass all test without the second `detectChanges`

```bash
Finished in 6 mins 35.946 secs / 6 mins 24.362 secs @ 13:40:33 GMT+0200 (Eastern European Standard Time)

SUMMARY:
✔ 3087 tests completed
ℹ 79 tests skipped
TOTAL: 3087 SUCCESS
```

```diff
--- a/src/clr-angular/wizard/wizard.spec.ts
+++ b/src/clr-angular/wizard/wizard.spec.ts
@@ -29,7 +29,6 @@ export default function(): void {
           wizard = context.clarityDirective;
           component = context.hostComponent;
           context.detectChanges();
-          context.detectChanges();
         });

         describe('open', () => {
```

File: src/clr-angular/forms/datalist/providers/datalist-id.service.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable } from '@angular/core';
+import { BehaviorSubject } from 'rxjs';
+import { Observable } from 'rxjs';
+
+let counter = 0;
+
+@Injectable()
+export class DatalistIdService {
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datalist/providers/datalist-id.service.ts]: This is doing the same thing as the ControlIdService. I would like to keep them as similar as possible.

File: src/clr-angular/forms/datalist/datalist-input.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    private datalistIdService: DatalistIdService
+  ) {
+    super(vcr, ClrDatalistContainer, injector, control, renderer, el);
+
+    if (!this.focusService) {
+      throw new Error('clrDatalist requires being wrapped in <clr-datalist-container>');
+    }
+  }
+
+  listValue;
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clr-angular/forms/datalist/datalist-input.ts]: By default, all members of a class in TypeScript are public.
I will add the type for it.

File: src/clr-angular/forms/tests/control.spec.ts (hunk 1 above)
Comment (hippee-lee) [reply src/clr-angular/forms/tests/control.spec.ts]: Thanks, these need to be tested to make sure they are present.

File: src/clr-angular/forms/datalist/datalist.ts (hunk 2 above)
Comment (hippee-lee) [reply src/clr-angular/forms/datalist/datalist.ts]: I will add the type.

File: src/clr-angular/wizard/wizard.spec.ts (hunk 3 above)
Comment (hippee-lee) [reply src/clr-angular/wizard/wizard.spec.ts]: I'm not sure how this change got in there, its not related to this PR. I wIll remove.

File: src/website/src/app/documentation/demos/datalist/datalist.demo.html [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                options that the user will need to filter or narrow down before making a choice. If there are 3-13
+                options, consider using the <a routerLink="../../select">select</a> input.
+            </p>
+
+            <h3>Auto-completion</h3>
+
+            <p>
+                Autocomplete is a familiar pattern for users of web technology. Searching, item suggestion and
+                filtering are all common activities users encounter when interacting with web technology and content.
+                The HTML5 element <code class="clr-code">dataliste</code> natively supports autocomplete behavior on
```
Comment (bdryanovski) [inline src/website/src/app/documentation/demos/datalist/datalist.demo.html]: there is a typo `datalist`

Comment (hippee-lee) [reply src/website/src/app/documentation/demos/datalist/datalist.demo.html]: This seems to be related to differences between vertical/horizontal form layouts. I'm looking into it.

File: src/clr-angular/forms/datalist/providers/datalist-id.service.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable } from '@angular/core';
+import { BehaviorSubject } from 'rxjs';
```
Comment (gnomeontherun) [inline src/clr-angular/forms/datalist/providers/datalist-id.service.ts]: nit: can be one import

File: src/clr-angular/forms/datalist/index.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+export * from './datalist-input';
```
Comment (gnomeontherun) [inline src/clr-angular/forms/datalist/index.ts]: Missing `datalist.ts`

File: src/clr-angular/forms/datalist/datalist.spec.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { DatalistIdService } from './providers/datalist-id.service';
+import { Component, Renderer2 } from '@angular/core';
+import { spec, TestContext } from '../../utils/testing/helpers.spec';
+import { ClrDatalist } from './datalist';
+import { ClrPopoverOpenCloseButton } from '../../utils/popover/popover-open-close-button';
+import { ClrPopoverEventsService, ClrPopoverPositionService, ClrPopoverToggleService } from '@clr/angular';
```
Comment (gnomeontherun) [inline src/clr-angular/forms/datalist/datalist.spec.ts]: Wrong import path

File: src/clr-angular/forms/datalist/datalist-input.spec.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        expect(() => {
+          const fixture = TestBed.createComponent(InvalidUseTest);
+          fixture.detectChanges();
+        }).toThrow();
+      });
+    });
+    TemplateDrivenSpec(ClrDatalistContainer, ClrDatalistInput, TemplateDrivenTest, 'clr-input');
+    // ReactiveSpec(ClrDatalistContainer, ClrDatalistInput, ReactiveTest, 'clr-input');
+
+    xdescribe('set datalist type', () => {
```
Comment (gnomeontherun) [inline src/clr-angular/forms/datalist/datalist-input.spec.ts]: `xdescribe`

File: src/clr-angular/forms/datalist/datalist-input.spec.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+          declarations: [InvalidUseTest],
+        });
+        expect(() => {
+          const fixture = TestBed.createComponent(InvalidUseTest);
+          fixture.detectChanges();
+        }).toThrow();
+      });
+    });
+    TemplateDrivenSpec(ClrDatalistContainer, ClrDatalistInput, TemplateDrivenTest, 'clr-input');
+    // ReactiveSpec(ClrDatalistContainer, ClrDatalistInput, ReactiveTest, 'clr-input');
```
Comment (gnomeontherun) [inline src/clr-angular/forms/datalist/datalist-input.spec.ts]: Reactive is commented out

File: src/website/src/app/documentation/demos/datalist/datalist.demo.html (hunk 6 above)
Comment (hippee-lee) [reply src/website/src/app/documentation/demos/datalist/datalist.demo.html]: This happens in horizontal layout b/c some inputs need  the right margin for an icon (datepicker). Vertical layouts remove the right margin. 
I added a targeted select for `clr-input-group` 's that are inside of the `clr-datalist-container`.

File: src/clr-angular/forms/datalist/providers/datalist-id.service.ts (hunk 7 above)
Comment (hippee-lee) [reply src/clr-angular/forms/datalist/providers/datalist-id.service.ts]: fixed

File: src/clr-angular/forms/datalist/index.ts (hunk 8 above)
Comment (hippee-lee) [reply src/clr-angular/forms/datalist/index.ts]: added

File: src/clr-angular/forms/datalist/datalist.spec.ts (hunk 9 above)
Comment (hippee-lee) [reply src/clr-angular/forms/datalist/datalist.spec.ts]: These were removed (unused imports).

File: src/clr-angular/forms/datalist/datalist-input.spec.ts (hunk 10 above)
Comment (hippee-lee) [reply src/clr-angular/forms/datalist/datalist-input.spec.ts]: I un x'ed it.

File: src/clr-angular/forms/datalist/datalist-input.spec.ts (hunk 11 above)
Comment (hippee-lee) [reply src/clr-angular/forms/datalist/datalist-input.spec.ts]: un commented Reactive.

File: src/website/src/app/documentation/demos/datalist/datalist.demo.html (hunk 6 above)
Comment (bdryanovski) [reply src/website/src/app/documentation/demos/datalist/datalist.demo.html]: I have almost the same issue with the datepicker additional clear icon, Maybe we need to rework this icon and input ordering pattern. 

I will open the conversation thread cause the typo is still there
```html
<code class="clr-code">dataliste</code>
```

**Comment by hippee-lee** [comment]
> I have almost the same issue with the datepicker additional clear icon, Maybe we need to rework this icon and input ordering pattern.
> 
> I will open the conversation thread cause the typo is still there
> 
> ```
> <code class="clr-code">dataliste</code>
> ```

Not sure how the typo did not get fixed. I just pushed the fix again. 
For the datalist dropdown icon - this is added by the browser, its not something we control. The scss changes are also specific to input groups inside the `clr-datalist-container`. It doesn't seem like this code can be easily generalized in this PR without re-working the `clr-input-group` and possibly other form input layout styles. I recommend keeping them separate for now until we can see a clear pattern that can be generalized.

**Comment by bdryanovski** [comment]
It seems that this issue with an additional icon inside this input-group is a lot more tricky than it looks. 

So we need to review and rework it - but agree that this is not part of the scope right now. At the same time, this is introducing visual bugs. 

So I'm giving a green light on this, but we need to track this issue and find a solution to that. There are a few feature requests that will need similar additional icons.

**Pull request merged**